### ISSUE TITLE
Typed Promise.all

### ISSUE THREAD

**Issue by haxiomic** [HaxeFoundation/haxe#10246]
Currently `Promise.all()` always gives you an `Array<Dynamic>`.

If the input was `Array<Promise<Type>>` (as it often is for `all()`), it would be nice to get `Array<Type>` in the callback

js.lib.Promise.all is declared as
```haxe
static function all(iterable:Array<Dynamic>):Promise<Array<Dynamic>>;
```

I think it would be better as
```haxe
static function all<T>(iterable:Array<Promise<T>>):Promise<Array<T>>;
```

If you have an array of promises of different types, then you'll have an `Array<Promise<Dynamic>>` so you get `Array<Dynamic>` out anyway

PR #10247

**Comment by kLabz** [comment]
> If you have an array of promises of different types, then you'll have an Array<Promise<Dynamic>> so you get Array<Dynamic> out anyway

But it wouldn't work without you explicitly typing as `Array<Promise<Dynamic>>`, right?

**Comment by haxiomic** [comment]
If you try to create an mixed array of promises: `[Promise.resolve(4), Promise.resolve('hello')]`, the compiler says `Arrays of mixed types are only allowed if the type is forced to Array<Dynamic>`

So I guess we want `Array<Dynamic>` as an overload
```haxe
@:overload(function(iterable:Array<Dynamic>):Promise<Array<Dynamic>> {})
static function all<T>(iterable:Array<Promise<T>>):Promise<Array<T>>;
```

The following mix and not-mixed now work:
```haxe
Promise.all([Promise.resolve(4), Promise.resolve('hello')]).then(a -> {
	$type(a); // Array<Dynamic>
});
```

```haxe
Promise.all([Promise.resolve(4), Promise.resolve(5)]).then(a -> {
	$type(a); // Array<Int>
});
```

**Comment by lublak** [comment]
Dynamic is rather a bad design, isn't it? But then it will be a breaking change.

**Comment by haxiomic** [comment]
Good point @lublak, and looking more at the Promise spec, turns out non-promise values are allowed, that is `Promise.all([1,2,3])` returns a promise to `[1,2,3]`, so I think then we need 1 more overload for this case

```haxe
@:overload(function(iterable:Array<Dynamic>):Promise<Array<Dynamic>> {})
@:overload(function<T>(iterable:Array<T>):Promise<Array<T>> {})
static function all<T>(iterable:Array<Promise<T>>):Promise<Array<T>>;
```

But the `Array<Dynamic>` seems to override the `Array<T>` so that one should be removed :(
```haxe
@:overload(function(iterable:Array<Dynamic>):Promise<Array<Dynamic>> {})
static function all<T>(iterable:Array<Promise<T>>):Promise<Array<T>>;
```

Now we have

```haxe
Promise.all([Promise.resolve(4), Promise.resolve('hello')]).then(a -> {
	$type(a); // Array<Dynamic>
});

Promise.all([Promise.resolve(4), Promise.resolve(5)]).then(a -> {
	$type(a); // Array<Int>
});

Promise.all([1, 2, 3]).then(a -> {
	$type(a); // Array<Dynamic>, not ideal but not worse than before
});

Promise.all([1, 2, 'three']).then(a -> {
	$type(a); // Array<Dynamic>
});
```

**Comment by lublak** [comment]
What if we add another overload with Array<T>? It was an idea but not tested. For [1, 2, 3]

**Comment by haxiomic** [comment]
I tried swapping the order of `Array<T>` and `Array<Dynamic>` overloads to change priority:

```haxe
@:overload(function<T>(iterable:Array<T>):Promise<Array<T>> {})
@:overload(function(iterable:Array<Dynamic>):Promise<Array<Dynamic>> {})
static function all<T>(iterable:Array<Promise<T>>):Promise<Array<T>>;
```

And it's close, it gets the types right but we get compiler errors so it's a breaking change:
```haxe
Promise.all([Promise.resolve(4), Promise.resolve('hello')]).then(a -> { // ! Arrays of mixed types are only allowed if the type is forced to Array<Dynamic>
	$type(a); // Array<Dynamic>
});

Promise.all([Promise.resolve(4), Promise.resolve(5)]).then(a -> {
	$type(a); // Array<Int>
});

Promise.all([1, 2, 3]).then(a -> {
	$type(a); // Array<Int>
});

Promise.all([1, 2, 'three']).then(a -> { // ! Arrays of mixed types are only allowed if the type is forced to Array<Dynamic>
	$type(a); // Array<Dynamic>
});
```

Maybe this is a bug on its own to be fixed, but I think on balance, given Promise.all on non-promises is likely to be rare in the wild we can do without solving that one right now and supporting `Array<Promise<T>>` is a step forward

**Comment by haxiomic** [comment]
Resolved in #10247
